autoload/vital/__vital__/Stream.vim

```
@@ -0,0 +1,368 @@
+    endwhile
+    if !open
+      let self.__end = 1
+    endif
+    return [list, open]
+  endfunction
+  function! stream.__estimate_size__() abort
+    return self._upstream.__estimate_size__()
+  endfunction
+  return stream
+endfunction
+
+function! s:Stream.limit(n) abort
+  let stream = deepcopy(s:Stream)
+  let stream._characteristics = s:ORDERED + s:SIZED + s:IMMUTABLE
+  let stream._upstream = self
+  let stream.__end = 0
+  let stream._n = max([a:n, 0])
+  function! stream.__take_possible__(...) abort
+    if self.__end
+      throw 'vital: Stream: stream has already been operated upon or closed'
+    endif
+    let list = self._n > 0 ? self._upstream.__take_possible__(self._n)[0] : []
+    let self.__end = (self.__estimate_size__() == 0)
+    return [list, !self.__end]
+  endfunction
+  function! stream.__estimate_size__() abort
+    return min([self._n, self._upstream.__estimate_size__()])
+  endfunction
+  return stream
+endfunction
+
+function! s:Stream.zip(stream) abort
+  return s:zip(self, a:stream)
+endfunction
+
+function! s:Stream.concat(stream) abort
+  return s:concat(self, a:stream)
+endfunction
+
+function! s:Stream.reduce(f, ...) abort
+  let Result = get(a:000, 0, 0)
+  let type = type(a:f)
+  if type is s:t_string
+    for Value in self.__take_possible__(self.__estimate_size__())[0]
+      let Result = map([[Result, Value]], a:f)[0]
+    endfor
+  elseif type is s:t_func
+    for Value in self.__take_possible__(self.__estimate_size__())[0]
+      let Result = a:f(Result, Value)
+    endfor
+  else
+    throw 'vital: Stream: reduce(): invalid type argument was given (Funcref or String or Data.Closure)'
+  endif
+  return Result
+endfunction
+
+function! s:Stream.max(...) abort
+  if self.__estimate_size__() == 0
+    return get(a:000, 0, 0)
+  endif
+  return self.reduce('max(v:val)', -1/0)
+endfunction
+
+function! s:Stream.max_by(f, ...) abort
+  if self.__estimate_size__() == 0
+    return get(a:000, 0, 0)
```

If the stream is empty and there is no default value, I feel we should raise an exception instead of `0`.